### PR TITLE
Adding notebook for MNIST training using PyTorch and StepFunctions

### PR DESCRIPTION
**Pull request by Avs163**
*Issue #, if available:* Adding notebook and MNIST.py 

*Description of changes:* Adding a notebook to create a simple workflow that involves a training step, creating a model, configuring an endpoint, and deploying the model. The notebook is to run Machine Learning pipelines on Step Functions, via MNIST. The folder originally existed in this same repository but got deleted and hence creating a PR to merge it back.

*Testing done:*

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

- [x] I have read the [CONTRIBUTING](https://github.com/aws/amazon-sagemaker-examples/blob/master/CONTRIBUTING.md) doc and adhered to the example notebook best practices
- [] I have updated any necessary documentation, including [READMEs](https://github.com/aws/amazon-sagemaker-examples/blob/master/README.md)
- [x] I have tested my notebook(s) and ensured it runs end-to-end
- [ ] I have linted my notebook(s) and code using `black-nb -l 100 {path}/{notebook-name}.ipynb`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: step-functions-data-science-sdk/training_pipeline_pytorch_mnist/code/mnist.py [hunk 1]
```diff
@@ -0,0 +1,260 @@
+import argparse
+import json
+import logging
+import os
+import sys
+
+import sagemaker_containers
+import torch
```
[the remark this answers is not in the document]
Comment (Avs163) [reply step-functions-data-science-sdk/training_pipeline_pytorch_mnist/code/mnist.py:8]: # Set a fixed random seed for reproducibility

SEED = 42
torch.manual_seed(SEED)
np.random.seed(SEED)
random.seed(SEED)

File: step-functions-data-science-sdk/training_pipeline_pytorch_mnist/code/mnist.py [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+        # multi-machine multi-gpu case
+        model = torch.nn.parallel.DistributedDataParallel(model)
+    else:
+        # single-machine multi-gpu case or single-machine or multi-machine cpu case
+        model = torch.nn.DataParallel(model)
+
+    optimizer = optim.SGD(model.parameters(), lr=args.lr, momentum=args.momentum)
+
+    for epoch in range(1, args.epochs + 1):
+        model.train()
```
[the remark this answers is not in the document]
Comment (Avs163) [reply step-functions-data-science-sdk/training_pipeline_pytorch_mnist/code/mnist.py:155]: When I am calling the  test() function on line 176, it runs model.eval() on line 181 before running the test/validation loop.

File: step-functions-data-science-sdk/training_pipeline_pytorch_mnist/code/mnist.py [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+            test_loss, correct, len(test_loader.dataset), 100.0 * correct / len(test_loader.dataset)
+        )
+    )
+
+
+def model_fn(model_dir):
+    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+    model = torch.nn.DataParallel(Net())
+    with open(os.path.join(model_dir, "model.pth"), "rb") as f:
+        model.load_state_dict(torch.load(f))
```
[the remark this answers is not in the document]
Comment (Avs163) [reply step-functions-data-science-sdk/training_pipeline_pytorch_mnist/code/mnist.py:204]: When I am calling the test() function on line 176, it runs model.eval() on line 181 before running the test/validation loop.

**Review by Avs163 — COMMENTED** [review]
Made three changes as recommended by codeguru:

1. Added a random seed for reproducibility
2. Verified that model.eval() is called before testing loop executes
3. Replaced torch.no_grad(): with with torch.inference_mode():

**Review by Avs163 — COMMENTED** [review]
Replaced torch.no_grad(): with with torch.inference_mode():.

**Comment by Avs163** [comment]
> example-pr logs says badge is missing. https://github.com/aws/amazon-sagemaker-examples/blob/main/CONTRIBUTING.md Here is the guide to include CI badge

I have added CI badge in the code. Please let me know if any other changes need to be made

**Comment by Avs163** [comment]
Reformatted the code, and now all required checks are passed.

**Review by zhaoqizqwang — APPROVED** [review]
Skipping grammar check
